### PR TITLE
refactor: accept `NamespaceName` instead of `&str` with `NamespaceRepo::create`

### LINKED ISSUE #7592: Catalog should use NamespaceName when creating namespaces
The catalog `NamespaceRepo::create()` accepts a bare `&str`:

https://github.com/influxdata/influxdb_iox/blob/d7dc3059728f8dc129eb8266fa5be39fe81a0b79/iox_catalog/src/interface.rs#L333-L339

This allows namespaces of arbitrary names to be created, bypassing namespace name validation.

Instead this call should take `name: &NamespaceName`. The type system upholds the invariant that any `NamespaceName` instance contains a valid namespace name. By accepting this instead of `&str`, the catalog can transitively uphold the same invariant - all namespaces created through the `NamespaceRepo::create()` use valid names.

### PR DESCRIPTION
**Pull request by kayagokalp**
Closes #7592.

`NamespaceRepo` trait's `create` function is changed from accepting a `&str` to `&NamespaceName` for the name of the new namespace to be created. This way it is not possible to create a namespace with an invalid name.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: router/src/namespace_resolver/ns_autocreation.rs [hunk 1]
```diff
@@ -107,7 +107,16 @@ where
                         .repositories()
                         .await
                         .namespaces()
-                        .create(namespace.as_str(), retention_period_ns)
+                        .create(
+                            &NamespaceName::new(namespace.as_str()).map_err(|_| {
```
Comment (alamb) [inline router/src/namespace_resolver/ns_autocreation.rs:111]: I am not sure if there is a cleaner way to create this error. Maybe @domodwyer  has some ideas

**Review by alamb — APPROVED** [review]
Thank you very much for the contribution @kayagokalp  -- I went through this PR carefully and it looks good to me. Sorry for the delay in review

@domodwyer  given you filed the request and this affects code on the router side, perhaps you would like to review as well

**Review by domodwyer — COMMENTED** [review]
Beautiful! 👌 

I love using types to enforce invariants, and this is a great example - had we been requiring a `NamespaceName` in the `create()` method before, it would have saved us from a bug in production! Fortunately it wasn't too bad at the time, but now it's impossible 🎉 

THANK YOU @kayagokalp ❤️ 

(small change to apply, then good to merge 🚀)

File: iox_catalog/src/interface.rs [hunk 2]
```diff
@@ -35,6 +35,9 @@ pub enum CasFailure<T> {
 #[allow(missing_copy_implementations, missing_docs)]
 #[snafu(visibility(pub(crate)))]
 pub enum Error {
+    #[snafu(display("invalid name: {}", name))]
```
[the remark this answers is not in the document]
Comment (kayagokalp) [reply iox_catalog/src/interface.rs:38]: Yeah I tried to keep them consistent with others, if you have a suggestion I can apply

[the remark this answers is not in the document]
Comment (alamb) [reply iox_catalog/src/interface.rs:38]: I think consistency is the best strategy in this case

**Review by domodwyer — APPROVED** [review]
Amazing! Thank you @kayagokalp 🎉

**Comment by domodwyer** [comment]
(pinging @wiedld - this is something you wanted to see added I believe!)

**Pull request merged**